tests/unit/test_charm.py

```diff
@@ -0,0 +1,160 @@
+        # Check status is Active
+        self.assertEqual(self.harness.charm.unit.status, ActiveStatus())
+
+    @patch("charm.TrainingOperatorCharm._create_crds")
+    @patch("charm.TrainingOperatorCharm._create_resource")
+    def test_blocked_on_appierror_on_install_event(self, create_auth, create_crds):
+        # Ensure the unit is in BlockedStatus
+        # on exception when creating auth resources
+        subtests = (
+            (
+                create_auth,
+                "Test BlockedStatus when ApiError is raised trying to create auth resources",
+            ),
+            (
+                create_crds,
+                "Test BlockedStatus when ApiError is raised trying to create CRDS",
+            ),
+        )
+        for create_type, subtest_description in subtests:
+            with self.subTest(msg=subtest_description):
+                create_type.side_effect = _FakeApiError()
+                try:
+                    self.harness.charm.on.install.emit()
+                except ApiError:
+                    self.assertEqual(
+                        self.harness.charm.unit.status,
+                        BlockedStatus(
+                            f"Creating/patching resources failed with code {create_type.side_effect.response.code}."
+                        ),
+                    )
+
+    @patch("charm.TrainingOperatorCharm._update_layer")
+    @patch("charm.TrainingOperatorCharm._patch_resource")
+    def test_config_changed_event(self, patch, update):
+        self.harness.charm.on.config_changed.emit()
+
+        # Ensure _update_layer is called once and
+        # _patch_resource is called
+        update.assert_called_once()
+        patch.assert_called()
+
+        # Ensure the unit is in BlockedStatus
+        # on exception when patching auth resources
+        patch.side_effect = _FakeApiError()
+        try:
+            self.harness.charm.on.config_changed.emit()
+        except ApiError:
+            self.assertEqual(
+                self.harness.charm.unit.status,
+                BlockedStatus(
+                    f"Patching resources failed with code {patch.side_effect.response.code}."
+                ),
+            )
+
+    @patch("charm.TrainingOperatorCharm._update_layer")
+    def test_on_training_operator_pebble_ready(self, update):
+        self.harness.container_pebble_ready("training-operator")
+
+        # Check the layer gets created
+        self.assertIsNotNone(
+            self.harness.get_container_pebble_plan("training-operator")._services
+        )
+
+    # Helpers
+    @patch("charm.Client.create")
+    def test_create_resource(self, client: MagicMock):
+        subtests = (
+            (
+                "auth",
+                {"namespace": "unit-kubeflow", "app_name": "unit-training-operator"},
+                "auth_manifests.yaml",
+            ),
+            ("crds", {}, "crds_manifests.yaml"),
+        )
+        for resource_type, context, template in subtests:
```

Comment:
I haven't used `unittest` over `pytest` in a while, but this section is IMO a point in favor of `pytest`. The `@pytest.mark.parametrize()` decorator is a lot nicer than this. I agree with the above conversation that we don't need to update this PR, just noting for future bikeshedding sessions :smile:

Reply:
I know :D @ca-scribner and I were having this conversation on Friday, we might want to move in the near future to pytest. Also in my queue of improvements.